shardy/dialect/sdy/ir/attrs.td

```diff
@@ -746,6 +798,19 @@ def Sdy_OpShardingRule : AttrDef<Sdy_Dialect, "OpShardingRule"> {
     these ops, so a user must tell it how. When it is a custom rule, then the
     rule is always preserved/never removed. `is_custom_rule` can only be true
     for `stablehlo.custom_call` ops.
+
+    **Constraints:**
+    - Number of operand/result mappings must match the number of
+      operands/results of the op.
+    - There is at least one mapping (can't have a rule for an op with no
+      operands/results).
+    - Rank of each `TensorMappingAttr` matches the rank of the corresponding
+      tensor type.
+    - Elements in `reduction_factors` and `need_replication_factors` must
```

Comment:
Any chance we can structure this as a group of constraints, something like:

For each group of factors (`reduction_factors`, `need_replication_factors`):
* Elements must be in range [0, `$factor_sizes`]
* No duplicate factor indices within group and across groups